core/include/engine/Pair_Sorting.hpp

```diff
@@ -0,0 +1,114 @@
+#ifndef SPIRIT_CORE_PAIR_SORTING_HPP
+#define SPIRIT_CORE_PAIR_SORTING_HPP
+#include "Spirit_Defines.h"
+#include <fmt/format.h>
+#include <algorithm>
+#include <engine/Vectormath_Defines.hpp>
+#include <numeric>
+
+namespace Engine
+{
+
+struct Pair_Order
+{
+    field<int> n_pairs_per_cell_atom;
+    field<int> offset_per_cell_atom;
+    std::vector<int> indices;
+
+    const field<Pair> const * pairs_ref;
+
+    Pair_Order() = default;
+
+    Pair_Order( field<Pair> & pairs, const field<int> & n_cells, int n_cell_atoms )
+            : n_pairs_per_cell_atom( field<int>( n_cell_atoms, 0 ) ),
+              offset_per_cell_atom( field<int>( n_cell_atoms, 0 ) ),
+              indices( std::vector<int>( pairs.size() ) ),
+              pairs_ref( &pairs )
+    {
+        std::iota( indices.begin(), indices.end(), 0 );
+
+        auto pair_compare_function = [&]( const int & idx_l, const int & idx_r ) -> bool {
+            const Pair & l = pairs[idx_l];
+            const Pair & r = pairs[idx_r];
+
+            if( l.i < r.i )
+                return true;
+            else if( l.i > r.i )
+                return false;
+
+            // sub-sorting
+            // if l.i == r.i, sort by the expected difference of the linear idx into the spins array
+            int d_idx_l
+                = l.j
+                  + n_cell_atoms
+                        * ( l.translations[0] + n_cells[0] * ( l.translations[1] + n_cells[1] * l.translations[2] ) );
+            int d_idx_r
+                = r.j
+                  + n_cell_atoms
+                        * ( r.translations[0] + n_cells[0] * ( r.translations[1] + n_cells[1] * r.translations[2] ) );
+
+            return d_idx_l < d_idx_r;
+        };
+
+        std::sort( indices.begin(), indices.end(), pair_compare_function );
+
+        // Count how many pairs interact with each cell atom
+        for( const auto & p : pairs )
+        {
+            n_pairs_per_cell_atom[p.i]++;
+        }
+
+        // Compute the offsets at which each cell atom has to look into the sorted pairs vector
+        for( int i = 1; i < n_pairs_per_cell_atom.size(); i++ )
+        {
+            offset_per_cell_atom[i] += offset_per_cell_atom[i - 1] + n_pairs_per_cell_atom[i - 1];
```

Comment:
Do I understand correctly that the resulting intervals of indices into the list of pairs **do not** overlap, because for the parallel Hamiltonian implementations the lists include the redundant pairs?